fix(#3878): WIP [Accessibility] Add tabindex, focusin/focusout, focus method and examples using focus management

majornista opened the pull request:
Work in progress to address #3878, by adding support for keyboard interaction an navigation of the scenegraph when rendering using SVG.

Add support for and examples using:
- tabindex,
- focusin/focusout event handlers,
- focus(item) event function,
- ariaRole and ariaRoleDescription on the View, so that an interactive visualization can have role="application."

New interactive stacked bar chart, attempting to recreate [an example from Data Navigator](https://dig.cmu.edu/data-navigator/#main), using roving tabindex and keyboard event handling to switch between navigating by column, and by stacked bar segment: https://github.com/vega/vega/blob/5e33eb2bb0cad32bd9833e789a42fce979eb5606/docs/examples/column-stacked-interactive.vg.json.

Modified Bar Chart and Airport Connections examples to support roving tab index for navigation of data points: 
- https://github.com/vega/vega/pull/3968/files#diff-2a6ca2cb2a1b7df0267b972fe076b93343484d425cb9b753d63eef803e47e459
- https://github.com/vega/vega/pull/3968/files#diff-216beca1f9861d45be7d644a77abcbebbe52ffff9a7d957abefa7c242c70c206

Interactive Line Chart example with navigable lines and points:
- https://github.com/vega/vega/blob/ddd72038d6480eec6db567e1a55d403e7f1ca0a5/docs/examples/interactive-line-chart.vg.json

Commit:
fix(#3878): [Accessibility] Device Dependent handlers are used DO NOT MERGE

- tabindex attribute on marks
- focusin/focusout event handling
- Adds support for focus(element) event method, so that one can move focus to SVGElements within the scenegraph.

Note that in order to tab in and out of the Vega view, we must allow keydown events in the Vega config.

Commit:
fix(#3878): [Accessibility] tabindex should respect interactive = false for marks

No tabindex attribute should be included on marks where interactive === false.

Commit:
fix(#3878): focus event function should work on an item() or group()

Add WIP stacked bar chart from Data Navigator to see if we can reproduce the functionality using focus management in Vega.

Commit:
fix(#3878): add outline property and draw focus outlines for consistency

Different browsers seem to render focus outlines for elements within an svg inconsistently. The new Interactive Stacked Bar Chart Example demonstrates a focus outline rendered within Vega, using SVG. The outline property allows a user to set the outline style for a mark or group, for example "outline: none."

Commit:
fix(#3878): suppress keyboard focus on ownerSVGElement

Adds tabindex="-1" to ownerSVGElement when it contains other focusable elements, so that the SVG element itself does not become an additional tab stop.

Commit:
fix(#3878): hide focus outline when there is no focusedItem

Fixes issue in Interactive Stacked Bar Chart example where the focus rectangle remained visible, surrounding a point with no width or height, when the user tabs away from the chart.

Commit:
fix(#3878): suppress keyboard focus on ownerSVGElement

Commit:
fix(#3878): trigger tooltips on focus with example

Commit:
fix(#3878): fix title shift when svg tooltip renders in the example

Commit:
fix(#3878): remove encoding for visibility from example

Commit:
fix(#3873): add ariaRole and ariaRoleDescription to View

Adds ability to set role="application"  to the view so that windows based screen readers switch into "forms mode" or "application mode" when entering an interactive visualization.

majornista commented:
I'm not sure why Test / Node 21 is failing. The tests pass for me running locally.

domoritz commented:
Let's make sure this is actually your code. I saw weird test failures in https://github.com/vega/vega/pull/3967.

Commit:
fix(#3878): add Interactive Line Chart example

Adds interactive line chart example that demonstrates managing focus between lines and points, and displaying tooltips for points.

Commit:
Merge remote-tracking branch 'upstream/main' into Issue-3878-keyboard-a11y

Commit:
Merge remote-tracking branch 'upstream/main' into Issue-3878-keyboard-a11y
